File: Day-10: Recursion and Backtracking/6) Word Break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        int n = s.size();
        vector<int> dp(n+1, -1);
        return helper(0, s, wordDict, dp);
    }
    int helper(int id, string &s, vector<string> &wordDict, vector<int> &dp)
    {
        int n = s.size();
        if(id == n)
        {
            return 1;
        }
        if(dp[id] != -1)
        {
            return dp[id];
        }
        for(string word : wordDict)
        {
            int len = word.size();
            if(id + len <= n && s.substr(id, len) == word)
            {
                if(helper(id + len, s, wordDict, dp) == 1)
                {
                    dp[id] = 1;
                    return dp[id];
                }
            }
        }
        dp[id] = 0;
        return dp[id];
    }
};

File: Day-10: Recursion and Backtracking/6) Word Break.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        int n = s.length();
        int[] dp = new int[n + 1];
        Arrays.fill(dp, -1);
        return helper(0, s, wordDict, dp) == 1;
    }
    public int helper(int idx, String s, List<String> wordDict, int[] dp)
    {
        int n = s.length();
        if(idx == n)
        {
            return 1;
        }
        if(dp[idx] != -1)
        {
            return dp[idx];
        }
        for(String word : wordDict)
        {
            int len = word.length();
            if(idx + len <= n && s.substring(idx, idx + len).equals(word))
            {
                if(helper(idx + len, s, wordDict, dp) == 1)
                {
                    dp[idx] = 1;
                    return 1;
                }
            }
        }
        dp[idx] = 0;
        return 0;
    }
}
